### PR TITLE
drivers: pwm: rpi_pico: Configuring the divide ratio adaptively

### PR DESCRIPTION
**Pull request by soburi**
As with many other implementations, we should set the division ratio to a value that corresponds to the interval specified in the API.

If the `divider-int-0` or variations of these for each channel properties
are not specified, or if these is 0,
the driver dynamically configures the division ratio by specified cycles.

The driver will operate at the specified division ratio if a non-zero
value is specified for `divider-int-0`.
This is unchanged from previous behavior.

Please specify ``divider-int-0`` explicitly to make the same behavior as
before.

In addition, the default device tree properties related to the division
ratio have been removed.

### CODE REVIEW

File: dts/bindings/pwm/raspberrypi,pico-pwm.yaml [hunk 1]
```diff
@@ -16,88 +16,72 @@ properties:
 
   divider-int-0:
     type: int
-    default: 1
     description: |
       The integral part of the divider for pwm slice 0.
       This number should be in the range 1 - 255. Defaults
```
Comment (yonsch) [inline dts/bindings/pwm/raspberrypi,pico-pwm.yaml:21]: Does it still default to the reset value?

Comment (soburi) [reply dts/bindings/pwm/raspberrypi,pico-pwm.yaml:21]: Yes, it does. It is set in this part.

https://github.com/zephyrproject-rtos/zephyr/pull/78786/files?diff=unified&w=0#diff-eb856606ea8e48c50d21f12c8393f76f68686f686d66d8817262913e99aff87aR190-R192

**Review by yonsch — DISMISSED** [review]

File: samples/drivers/led/pwm/Kconfig [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config BLINK_DURATION_SHORT
+	int "LED on and off times for short-cycle blinking demo"
+	default 100
+	help
+	  Specifies the LED blinking time in milliseconds in the short-cycle
```
Comment (simonguinot) [inline samples/drivers/led/pwm/Kconfig:8]: And maybe: Specifies the LED on/off delay in milliseconds for short cycle blinking demonstration.

File: samples/drivers/led/pwm/Kconfig [hunk 3]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config BLINK_DURATION_SHORT
```
Comment (simonguinot) [inline samples/drivers/led/pwm/Kconfig:4]: Suggestion: `BLINK_DELAY_SHORT`

File: samples/drivers/led/pwm/Kconfig [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	  blinking demo will not be performed.
+
+config FADE_DELAY
+	int "delay time for fade demo"
+	default 10
+	help
+	  Specifies the delay time for the fade demo of the PWM-LED sample.
+	  The fade demo increases the brightness by 1 percent point every
+	  time this delay time elapses and ends when it reaches the value
+	  of MAX_BRIGHTNESS.
```
Comment (simonguinot) [inline samples/drivers/led/pwm/Kconfig:27]: Suggestion: The brightness gradually increases by one level each time this delay elapses.

Let's not talk about percent. When we will update the maximum brightness to 255, we might forget to update this comment :)

File: samples/drivers/led/pwm/Kconfig [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	help
+	  Specifies the LED blinking time in milliseconds in the long-cycle
+	  blinking demo of the PWM-LED sample. If set to 0, the long-cycle
+	  blinking demo will not be performed.
+
+config FADE_DELAY
+	int "delay time for fade demo"
+	default 10
+	help
+	  Specifies the delay time for the fade demo of the PWM-LED sample.
```
Comment (simonguinot) [inline samples/drivers/led/pwm/Kconfig:24]: Suggestion: Specifies the delay in milliseconds for ...

File: samples/drivers/led/pwm/src/main.c [hunk 6]
```diff
@@ -66,28 +63,38 @@ static void run_led_test(const struct device *led_pwm, uint8_t led)
...
-		k_sleep(FADE_DELAY);
+		k_sleep(K_MSEC(CONFIG_FADE_DELAY));
 	}
 	k_sleep(K_MSEC(1000));
 
-	/* Set LED blinking (on: 0.1 sec, off: 0.1 sec) */
-	err = led_blink(led_pwm, led, 100, 100);
+#if CONFIG_BLINK_DURATION_SHORT > 0
+	/* Start LED blinking (short cycle) */
+	err = led_blink(led_pwm, led, CONFIG_BLINK_DURATION_SHORT, CONFIG_BLINK_DURATION_SHORT);
```
Comment (simonguinot) [inline samples/drivers/led/pwm/src/main.c:72]: Optionally, limit the line length to 80 columns, to stay consistent with the rest of the code.

**Review by simonguinot — COMMENTED** [review]
Hi @soburi,

The LED part looks OK but I am not sure about the PWM driver. Please find some questions and suggestions below.

File: dts/bindings/pwm/raspberrypi,pico-pwm.yaml [hunk 7]
```diff
@@ -16,88 +16,72 @@ properties:
 
   divider-int-0:
```
[the remark this answers is not in the document]
Comment (soburi) [reply dts/bindings/pwm/raspberrypi,pico-pwm.yaml:17]: Your understanding is correct.
The current implementation is as such specs.
This PR intends to relax the limitation.

File: drivers/pwm/pwm_rpi_pico.c [hunk 8]
```diff
@@ -160,19 +187,23 @@ static int pwm_rpi_init(const struct device *dev)
...
 }
 
-#define PWM_INST_RPI_SLICE_DIVIDER(idx, n)				   \
-	{								   \
-		.integral = DT_INST_PROP(idx, UTIL_CAT(divider_int_, n)), \
-		.frac = DT_INST_PROP(idx, UTIL_CAT(divider_frac_, n)),	   \
+#define PWM_INST_RPI_SLICE_DIVIDER(idx, n)                                                         \
+	{                                                                                          \
+		.integral = DT_INST_PROP_OR(idx, UTIL_CAT(divider_int_, n), 0),                    \
+		.frac = DT_INST_PROP_OR(idx, UTIL_CAT(divider_frac_, n), 0),                       \
```
[the remark this answers is not in the document]
Comment (soburi) [reply drivers/pwm/pwm_rpi_pico.c:206]: This follows the clang-format convention.
I could also change the section below accordingly,
but this PR doesn't require any modifications, so we'll leave it as is.

File: samples/drivers/led/pwm/Kconfig [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config BLINK_DURATION_SHORT
+	int "LED on and off times for short-cycle blinking demo"
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/drivers/led/pwm/Kconfig]: Thank you for the suggestion. I applied it.

File: samples/drivers/led/pwm/Kconfig (hunk 3 above)
Comment (soburi) [reply samples/drivers/led/pwm/Kconfig]: I applied it.

File: samples/drivers/led/pwm/Kconfig (hunk 5 above)
Comment (soburi) [reply samples/drivers/led/pwm/Kconfig]: I applied it.

File: samples/drivers/led/pwm/Kconfig (hunk 4 above)
Comment (soburi) [reply samples/drivers/led/pwm/Kconfig]: I applied it.

File: dts/bindings/pwm/raspberrypi,pico-pwm.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (soburi) [reply dts/bindings/pwm/raspberrypi,pico-pwm.yaml]: I forgot to update the description.
The "automatic divider handling" behavior is extended in this PR.
I added a description of this feature.

File: samples/drivers/led/pwm/Kconfig (hunk 2 above)
Comment (soburi) [reply samples/drivers/led/pwm/Kconfig]: I applied it.

File: dts/bindings/pwm/raspberrypi,pico-pwm.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonguinot) [reply dts/bindings/pwm/raspberrypi,pico-pwm.yaml:17]: OK understood.

**Review by simonguinot — APPROVED** [review]
Hi @soburi,

I think that the PWM driver configuration is a bit confusing. And if it were me, I would have removed the `divisor-{int,frac}-` DT properties. But overall, it looks good. Thanks !

**Comment by simonguinot** [comment]
Hi @yonsch. Please can you revisit ?

**Comment by soburi** [comment]
@yonsch Could you re-approve this please?

**Review by yonsch — APPROVED** [review]

**Pull request merged**